### PR TITLE
Developement

### PR DESCRIPTION
**Pull request by nevaldas1**
Implemented calculator

### CODE REVIEW

File: src/main/java/com/altf4/app/controller/LoanCalculatorController.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+@RequiredArgsConstructor
+public class LoanCalculatorController {
+
+    private LoanCalculatorService loanCalculatorService;
+
+    @Autowired
+    public LoanCalculatorController (LoanCalculatorService loanCalculatorService) {
+        this.loanCalculatorService = loanCalculatorService;
+    }
+
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/controller/LoanCalculatorController.java:21]: Remove blank space

File: src/main/java/com/altf4/app/controller/LoanCalculatorController.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package com.altf4.app.controller;
+
+import com.altf4.app.model.LoanCalculationResponse;
+import com.altf4.app.model.LoanCalculationRequest;
+import com.altf4.app.service.LoanCalculatorService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/controller/LoanCalculatorController.java:9]: Remove blank space

File: src/main/java/com/altf4/app/model/LoanCalculationResponse.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        public LoanCalculationResponseBuilder totalPaymentSum(int totalPaymentSum) {
+            this.totalPaymentSum = totalPaymentSum;
+            return this;
+        }
+
+        public LoanCalculationResponse build() {
+            return new LoanCalculationResponse(this);
+        }
+    }
+
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/model/LoanCalculationResponse.java]: Remove blank space

File: src/main/java/com/altf4/app/service/LoanCalculatorService.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                     / (Math.pow(1 + monthlyInterestRate, loanTermMonths) - 1) / 10.0) * 10);
+    }
+
+    private static LoanCalculationResponse buildResponse(LoanCalculationRequest request, int monthlyPayment) {
+        return new LoanCalculationResponse.LoanCalculationResponseBuilder()
+                .monthlyPaymentAmount(monthlyPayment)
+                .interestRate(EURIBOR_RATE + MARGIN)
+                .loanAmount(calculateLoanAmount(request))
+                .totalInterestAmount(calculateTotalInterestAmount(request, monthlyPayment))
+                .totalPaymentSum(calculateTotalPaymentSum(request, monthlyPayment))
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/service/LoanCalculatorService.java:51]: So why did you decide to leave it?

File: src/main/java/com/altf4/app/service/LoanCalculatorService.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    private static int calculateTotalInterestAmount(LoanCalculationRequest request, int monthlyPayment) {
+        return monthlyPayment * 12 * request.getTermYears() - (request.getTotalAmount() - request.getDownPayment());
+    }
+    private static int calculateTotalPaymentSum(LoanCalculationRequest request, int monthlyPayment) {
+        return calculateLoanAmount(request) + calculateTotalInterestAmount(request, monthlyPayment);
+    }
+
+
+
+
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/service/LoanCalculatorService.java]: Remove blank space

File: src/main/java/com/altf4/app/validator/LoanTermValidator.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import com.altf4.app.exception.ValidationException;
+
+public class LoanTermValidator extends AbstractValidator<Integer> {
+    @Override
+    public void validate(Integer termYears) throws ValidationException {
+
+        if (termYears > 30 || termYears < 1) {
+            throw new ValidationException("Requested loan term must be between 1 and 30 years.");
+        }
+
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/validator/LoanTermValidator.java]: Remove blank space

File: src/main/java/com/altf4/app/validator/MinimalLoanValueValidator.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+public class MinimalLoanValueValidator extends AbstractValidator <Integer> {
+
+    @Override
+    public void validate(Integer totalAmount) throws ValidationException {
+
+        if (totalAmount < 10000) {
+            throw new ValidationException("Requested loan must be at least 10.000 euros.");
+        }
+
```
Comment (NeHerkus) [inline src/main/java/com/altf4/app/validator/MinimalLoanValueValidator.java]: Remove blank space

**Review by NeHerkus — CHANGES_REQUESTED** [review]
Nicely done, just some minor things to fix! 👍

File: src/main/java/com/altf4/app/service/LoanCalculatorService.java (hunk 4 above)
Comment (nevaldas1) [reply src/main/java/com/altf4/app/service/LoanCalculatorService.java:51]: We had an agreement with Sharat to leave all maths logic in the back-end. If you don't support such decision, I'll delete it.

Comment (NeHerkus) [reply src/main/java/com/altf4/app/service/LoanCalculatorService.java:51]: Well adding two numbers is not that complex of a logic 😅
But sure, let's imagine, that there might be some advanced things later on.

**Review by NeHerkus — APPROVED** [review]

**Review by NeHerkus — APPROVED** [review]

**Review by bridzius — APPROVED** [review]

**Pull request merged**